HIP/hip-844.md

```diff
@@ -0,0 +1,86 @@
+
+## Motivation
+
+Unclear rules for when the nonce of an EthereumTransaction signer account should be updated result in discrepancies between the Consensus nodes' and Mirror nodes' states. This leads to users experiencing issues when reading the account nonce from the Mirror node and trying to use it in a transaction sent to the Consensus node.
+
+## Rationale
+
+The nonce of a signer of an Ethereum transaction should be increased for all transactions that are executed in the EVM.
+
+Transactions that have reached consensus and are handled by the node but have failed additional validations performed by the node prior to being executed in the EVM should have no effect on the Ethereum transaction signer nonce.
+
+A new protobuf field in the record stream - signerNonce - should be populated by the node with the current value of the signer account nonce after executing the transaction. This value should be used by the Mirror nodes when importing record stream files.
+
+## User stories
+
+1. As a Mirror node operator, I would like to receive information about incremented account nonces from the record stream produced by the consensus nodes.
+2. As a web3 user of the network, I would like to be able to more easily submit transactions without having to change the transaction nonce manualy.
+
+## Specification
+
+### Protobuf update
+
+Add a new protobuf field to contract_call_local.proto:
+
+```
+message ContractFunctionResult {
+    
+    [...]
+
+    /**
+     * If not null this field specifies what the value of the signer account nonce is post transaction execution. 
+     * For transactions that don't update the signer nonce, this field should be null.
+     */
+    Int64Value signer_nonce = 15;
+}
+```
+
+### Services update
+
+1. Move the signer nonce update to be right before this code block that starts the transaction execution in the EVM:
+```
+while (!messageFrameStack.isEmpty()) {
+   process(messageFrameStack.peekFirst(), tracer);
+}
+```
+2. Populate the new protobuf field when externalising the transaction.
+
+### Mirror node update
+
+When importing transactions from the record stream, update the value of the signer account's nonce that's persisted on the mirror node side with the value externalised in the new protobuf field.
+
+## Backward Compatibility
+
+When importing legacy transactions that don't have this field, the Mirror node should consider whether the transaction result is SUCCESS or CONTRACT_REVERT_EXECUTED and increment the signer nonce in these cases. Some statuses such as INSUFFICIENT_GAS can be returned both before and after entering the EVM, there is no way for the Mirror node to know whether the signer nonce should be updated. This was the root cause of multiple issues and the reason for the new protobuf field.
```

Comment:
Will there be any consideration for a sidecar based nonce migration? Or are customs okay with it only being accurate going forward?

Reply:
There is no migration planned at this stage.

Reply:
Perhaps add a note to the doc somewhere about no migration and will only provide accurate nonces for new accounts.

Reply:
added note